### PR TITLE
Make it easier to add new usage metrics.

### PR DESCRIPTION
**Pull request by grokys**
Currently, adding a new usage metric is a bit of a PITA. You need to:

- Add a new field to `UsageModel`
- Add a new method to `IUsageTracker`
- Add a new method to `UsageTrackerDispatcher`
- Add a new method to `UsageTracker`
- Add the field to `UsageTracker.ClearCounters`
- Add the field to `UsageModel.Clone`

This PR:

- [x] Removes the individual `IncrementX` methods from `IUsageTracker` and replaces them with an `IncrementCounter` method which takes a lambda describing the metric to update.
- [x] Uses reflection to reset the counters in `UsageTracker.ClearCounters`
- [x] Uses reflection to clone the counters in `UsageModel.Clone`
- [x] Moves IO and date comparison operations out of `UsageTracker` so it can be unit tested
- [x] Adds unit tests for `UsageTracker`

### CODE REVIEW

File: src/GitHub.VisualStudio/Services/UsageService.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            using (var s = File.OpenRead(path))
+            using (var r = new StreamReader(s, Encoding.UTF8))
+            {
+                return await r.ReadToEndAsync();
+            }
+        }
+
+        async Task WriteAllTextAsync(string path, string text)
+        {
+            using (var s = File.OpenWrite(path))
```
Comment (jcansdale) [inline src/GitHub.VisualStudio/Services/UsageService.cs]: We should be creating a new file, not overwriting the old one.

File: src/GitHub.VisualStudio/GitHub.VisualStudio.csproj [hunk 2]
```diff
@@ -11,6 +11,7 @@
     <NuGetPackageImportStamp>
     </NuGetPackageImportStamp>
     <TargetFrameworkProfile />
+    <ApplicationVersion>2.3.1.0</ApplicationVersion>
```
Comment (shana) [inline src/GitHub.VisualStudio/GitHub.VisualStudio.csproj]: This change shouldn't be in here. Bad merge somewhere?

File: src/GitHub.VisualStudio/Services/UsageService.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                storePath = Path.Combine(
+                    Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
+                    program.ApplicationName,
+                    StoreFileName);
+            }
+        }
+
+        async Task<string> ReadAllTextAsync(string path)
+        {
+            using (var s = File.OpenRead(path))
```
Comment (shana) [inline src/GitHub.VisualStudio/Services/UsageService.cs]: You can use rothko for IO services that can be mocked for unit testing

File: src/GitHub.VisualStudio/Services/UsageTracker.cs [hunk 4]
```diff
@@ -1,319 +1,105 @@
...
-        }
-
         async Task TimerTick()
         {
             await Initialize();
 
-            if (firstRun)
+            if (firstTick)
             {
                 await IncrementLaunchCount();
```
Comment (shana) [inline src/GitHub.VisualStudio/Services/UsageTracker.cs]: `IncrementLaunchCount` calls `LoadUsage` which calls `await Initialize` which has already been called right before this. Do we need these double `await Initialize` calls?

File: src/GitHub.VisualStudio/Services/UsageTracker.cs [hunk 5]
```diff
@@ -1,319 +1,105 @@
...
                 connectionManager = gitHubServiceProvider.GetService<IConnectionManager>();
                 userSettings = gitHubServiceProvider.GetService<IPackageSettings>();
                 vsservices = gitHubServiceProvider.GetService<IVSServices>();
-
-                var program = gitHubServiceProvider.GetService<IProgram>();
-                storePath = System.IO.Path.Combine(
-                    Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
-                    program.ApplicationName,
-                    StoreFileName);
+                initialized = true;
```
Comment (shana) [inline src/GitHub.VisualStudio/Services/UsageTracker.cs]: Is there a possibility of a race condition here where `await Initialize` gets called at the same time?

File: src/GitHub.VisualStudio/Services/UsageService.cs (hunk 3 above)
Comment (grokys) [reply src/GitHub.VisualStudio/Services/UsageService.cs]: Already thought of that, but Rothko doesn't have the methods we need as far as I can tell, which is partly why I split this I/O functionality out into a separate service apart from the usage tracker itself.

File: src/GitHub.VisualStudio/GitHub.VisualStudio.csproj (hunk 2 above)
Comment (grokys) [reply src/GitHub.VisualStudio/GitHub.VisualStudio.csproj]: Yep, bad merge - well spotted!

File: src/GitHub.VisualStudio/Services/UsageTracker.cs (hunk 5 above)
Comment (grokys) [reply src/GitHub.VisualStudio/Services/UsageTracker.cs]: There is a small possibility, if a counter is initialized on a background thread and while switching to the main thread, another counter is initialized. However all that would happen in that scenario is that services would get loaded twice, which I didn't think was a big enough problem to warrant putting a `SemaphoreSlim` or similar around.

File: src/GitHub.VisualStudio/Services/UsageTracker.cs (hunk 4 above)
Comment (grokys) [reply src/GitHub.VisualStudio/Services/UsageTracker.cs]: Actually, yes we can remove the `Initialize` at the top of `TimerTick`! 👍

**Comment by grokys** [comment]
@shana any chance of another review of this? I seem to remember you expressing surprise that this hadn't already gone in a few weeks back.

**Pull request merged**